docs/source/rfc/isa_restructuring.org

```diff
@@ -0,0 +1,732 @@
+instead of a dynamically loadable permission validator, there would be
+a dynamically loadable executor, which is responsible not only for
+checking if the instruction application violates any of the
+invariants, and checks for permissions, but is responsible wholly for
+upholding the blockchain invariants.
+
+As such, under the executor model, instructions are exposed as objects
+for which the application is a well-defined fallible operation.  The
+executor operates on a per-transaction basis, meaning that if a set of
+instructions is submitted, in principle, the executor can allow or
+deny the applications of certain instruction combinations.  At which
+point the permission tokens became ill-defined, as their meaning is
+undefined until a specific executor is loaded.  Within these
+constraints, the distinction between blockchain and host invariants
+becomes clearly defined: everything that isn't enforced by a failure
+in applying an instruction is a blockchain invariant, including
+permissions. The executor is responsible for enforcing the blockchain
+invariants, and enforcement of host instructions is propagated from
+the host to the executor as well.
+
+* Shortcomings of the current ISA
+
+** The Expression problem
+
+Firstly, while the Iroha blockchain ledger has an expression language
+that is dangerously close to being Turing complete (more precisely by
+being a lambda calculus), the expression language is largely unused.
+Any simple logic typically entails dealing with a plain instruction,
+which due to the limitations of the expression language cannot be
+strongly typed, and instead recovers the type through multiple nested
+tagged unions and a process of evaluation. The number of modes of
+failure for any given instruction is unnecessarily high, eschewing
+many of the advantages of statically and strictly typed languages.
+
+Of course, there is a trade-off for which it was made. An instruction
+can target something which is not known ahead of time, or before the
+transaction reaches the ledger. The expression language has facilities
+to substitute any query into any location. This can also be extended
+to sequences of instructions, and because the aforementioned
+instructions accept expression-valued arguments, this can be
+arbitrarily complex. However, by virtue of being interpreted, the
+expressions are slow. Due to having the user compose expressions and
+dealing with a threadbare model that does not have a front-end in the
+form of a DSL, the model isn't particularly expressive either.
+Further, to add insult to injury, while this model could in principle
+be extensible, given that expressions within /certain tags only/ are
+implemented, it would be possible to add custom, blockchain-specific
+expressions. However, at present this is not possible. As an example
+of where this functionality would have been useful, is CSD/RTGS
+linkages Proof of concept. In which an entire smart-contract had been
```

Comment:
Can we write this info in a public document?

Reply:
At this point the entire report is public, so yes. Before the end of my notice period I'll upload the report to the Iroha 2 docs repo, so we can actually refer to the precise locations where I point out what we would have found useful.